btcec/signature.go

```diff
@@ -37,43 +36,32 @@ var (
 	oneInitializer = []byte{0x01}
 )
 
-// Serialize returns the ECDSA signature in the more strict DER format.  Note
-// that the serialized bytes returned do not include the appended hash type
-// used in Bitcoin signature scripts.
-//
-// encoding/asn1 is broken so we hand roll this output:
-//
-// 0x30 <length> 0x02 <length r> r 0x02 <length s> s
+// Serialize returns the ECDSA signature as r || s,
+// where both r and s are encoded into 32 byte big endian integers.
+// In order to remove malleability,
+// we set s = curve_order - s, if s is greater than curve.Order() / 2.
 func (sig *Signature) Serialize() []byte {
 	// low 'S' malleability breaker
 	sigS := sig.S
 	if sigS.Cmp(S256().halfOrder) == 1 {
 		sigS = new(big.Int).Sub(S256().N, sigS)
 	}
-	// Ensure the encoded bytes for the r and s values are canonical and
-	// thus suitable for DER encoding.
-	rb := canonicalizeInt(sig.R)
-	sb := canonicalizeInt(sigS)
-
-	// total length of returned signature is 1 byte for each magic and
-	// length (6 total), plus lengths of r and s
-	length := 6 + len(rb) + len(sb)
-	b := make([]byte, length)
-
-	b[0] = 0x30
-	b[1] = byte(length - 2)
-	b[2] = 0x02
-	b[3] = byte(len(rb))
-	offset := copy(b[4:], rb) + 4
-	b[offset] = 0x02
-	b[offset+1] = byte(len(sb))
-	copy(b[offset+2:], sb)
-	return b
+	rBytes := sig.R.Bytes()
+	sBytes := sigS.Bytes()
+	sigBytes := make([]byte, 64)
+	// 0 pad the byte arrays from the left if they aren't big enough.
+	copy(sigBytes[32-len(rBytes):32], rBytes)
+	copy(sigBytes[64-len(sBytes):64], sBytes)
+	return sigBytes
 }
 
 // Verify calls ecdsa.Verify to verify the signature of hash using the public
-// key.  It returns true if the signature is valid, false otherwise.
+// key. It returns true if the signature is valid, false otherwise.
+// This fails on any signature with signatures in non-canonical form.
```

Comment:
Nit: s/This fails on any signature with signatures in non-canonical form./This fails on any signature in non-canonical form. or  This fails on any signature with signatures in non-canonical form.

Actually, if you ever expect this library to be used for existing signatures which do not follow the canonical form for signatures, it would be good to add a sentence here as well what the canonical form actually means. So far it only appears in the comment for `Serilaize`. If this will only be used to verify signatures coming from this lib, please ignore this comment (you can even delete the sentence as we only produce signatures in canonical form).

P.S. I found the documentation of the bitcoin's secp256 C implementation exemplary (no need to replicate this extensive documentation for now though):
https://github.com/bitcoin/bitcoin/blob/9ccafb1d7bdd172a9b963444072a844da379c4f7/src/secp256k1/include/secp256k1.h

Reply:
I think this should only be used to verify signatures coming from this lib, but added a comment anyway, as its good to be clear about whats going on :)